File: Problems/173_Binary_Search_Tree_Iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# using stack
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.stack = []
        self.curr = root

    def next(self) -> int:
        while self.curr:
            self.stack.append(self.curr)
            self.curr = self.curr.left
        self.curr = self.stack.pop()
        res = self.curr.val
        self.curr = self.curr.right
        return res

    def hasNext(self) -> bool:
        return self.stack or self.curr

# using yield
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.iterator = self.inorder(root)
        self.next_num = next(self.iterator, None)

    def next(self) -> int:
        curr = self.next_num
        self.next_num = next(self.iterator, None)
        return curr
        
    def hasNext(self) -> bool:
        return self.next_num is not None
    
    def inorder(self, node):
        if not node:
            return
        yield from self.inorder(node.left)
        yield node.val
        yield from self.inorder(node.right)


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: Problems/173_Binary_Search_Tree_Iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
public:
    TreeNode* curr;
    stack<TreeNode*> st;
    BSTIterator(TreeNode* root) {
        this->curr = root;
    }
    
    int next() {
        while(curr){
            st.push(curr);
            curr = curr->left;
        }
        
        curr = st.top();
        st.pop();
        int res = curr->val;
        curr = curr->right;
        return res;
    }
    
    bool hasNext() {
        return !st.empty() || curr;
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */